### PR TITLE
Add automatic cluster calculations

### PR DESCRIPTION
**Pull request by jteijema**
This PR adds automatic cluster calculation. It works, but I calculate K-means 1 time too often. It's a very light calculation so maybe its okay but I'd like to reduce it.

### CODE REVIEW

**Comment by jteijema** [comment]
Decided to do the optimization later if it turns out to take long. For now the embedding generation takes factors longer.

File: asreviewcontrib/semantic_clustering/semantic_clustering.py [hunk 1]
```diff
@@ -75,16 +77,50 @@ def SemanticClustering(asreview_data_object):
...
     # visualize clusters
     print("Visualizing clusters...")
     tsne_data = [tsne[:, 0], tsne[:, 1]]
     visualize_clusters(tsne_data, labels)
 
 
+# Optimal n clusters, very inefficient, to be corrected in a future PR
+def calc_optimal_n_clusters(features):
+
+    Sum_of_squared_distances = []
```
Comment (J535D165) [inline asreviewcontrib/semantic_clustering/semantic_clustering.py:112]: Lowercase please. See pep8 style guide.

**Comment by J535D165** [comment]
Nice PR, let's merge soon.

File: asreviewcontrib/semantic_clustering/semantic_clustering.py (hunk 1 above)
Comment (jteijema) [reply asreviewcontrib/semantic_clustering/semantic_clustering.py:112]: Visual studio automatically applies flake8 to my files. Let me see why this didn't adjust.

Comment (jteijema) [reply asreviewcontrib/semantic_clustering/semantic_clustering.py:112]: I will adjust this in a future PR

File: asreviewcontrib/semantic_clustering/semantic_clustering.py [hunk 2]
```diff
@@ -36,7 +38,7 @@ def SemanticClustering(asreview_data_object):
     data = load_data(asreview_data_object)
 
     # cut data for testing
```
Comment (Rensvandeschoot) [inline asreviewcontrib/semantic_clustering/semantic_clustering.py]: Can you explain why this is needed and why 30?

File: asreviewcontrib/semantic_clustering/semantic_clustering.py [hunk 3]
```diff
@@ -75,16 +77,50 @@ def SemanticClustering(asreview_data_object):
...
 
+    # calculate optimal number of clusters
+    print("Calculating optimal number of clusters...")
+    n_clusters = calc_optimal_n_clusters(tsne)
+    print("Optimal number of clusters: ", n_clusters)
+
     # run k-means
     print("Running k-means...")
-    labels = run_KMeans(tsne, 10, 10)
+    labels = run_KMeans(tsne, n_clusters, 10)
```
Comment (Rensvandeschoot) [inline asreviewcontrib/semantic_clustering/semantic_clustering.py:99]: why 10?

File: asreviewcontrib/semantic_clustering/semantic_clustering.py [hunk 4]
```diff
@@ -75,16 +77,50 @@ def SemanticClustering(asreview_data_object):
...
     tsne_data = [tsne[:, 0], tsne[:, 1]]
     visualize_clusters(tsne_data, labels)
 
 
+# Optimal n clusters, very inefficient, to be corrected in a future PR
+def calc_optimal_n_clusters(features):
+
+    Sum_of_squared_distances = []
+
+    K = range(1, 25)
```
Comment (Rensvandeschoot) [inline asreviewcontrib/semantic_clustering/semantic_clustering.py:114]: why 25? Should this be a setting that can be adjusted by the user?

**Review by Rensvandeschoot — CHANGES_REQUESTED** [review]

File: asreviewcontrib/semantic_clustering/semantic_clustering.py [hunk 5]
```diff
@@ -13,6 +13,8 @@
 # limitations under the License.
 
 # imports
+from sklearn.cluster import KMeans
```
Comment (J535D165) [inline asreviewcontrib/semantic_clustering/semantic_clustering.py]: Please group imports. See ASReview LAB

File: asreviewcontrib/semantic_clustering/semantic_clustering.py (hunk 3 above)
Comment (jteijema) [reply asreviewcontrib/semantic_clustering/semantic_clustering.py:99]: Well, I copied this from the original code. I will have to look this up!

File: asreviewcontrib/semantic_clustering/semantic_clustering.py (hunk 4 above)
Comment (jteijema) [reply asreviewcontrib/semantic_clustering/semantic_clustering.py:114]: No, this range is already way over the maximal 'comfortable' amount of clusters, and 25 is only there to limit the algorithm runs. KMeans is quite a light algorithm performance wise and so 25 could be set quite high.

**Comment by jteijema** [comment]
That was all, should be ready to merge.

**Review by Rensvandeschoot — APPROVED** [review]

**Pull request merged**